### PR TITLE
OpenSearch AWS autodiscovery

### LINKED ISSUE #26670: Support OpenSearch AutoDiscovery.
### What
Support Auto discovery for AWS OpenSearch.

### PR DESCRIPTION
**Pull request by Tener**
Closes https://github.com/gravitational/teleport/issues/26670

### CODE REVIEW

File: lib/services/database.go [hunk 1]
```diff
@@ -924,6 +925,90 @@ func newElastiCacheDatabase(cluster *elasticache.ReplicationGroup, endpoint *ela
...
+	var databases types.Databases
+
+	if aws.StringValue(domain.Endpoint) != "" {
+		metadata, err := MetadataFromOpenSearchDomain(domain, apiawsutils.OpenSearchDefaultEndpoint)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		meta := types.Metadata{
+			Description: fmt.Sprintf("OpenSearch domain in %v, default endpoint", metadata.Region),
```
Comment (greedy52) [inline lib/services/database.go:940]: nit: better to be consistent with other namings like `OpenSearch domain in %v` or `OpenSearch domain in %v (default endpoint)`

File: api/proto/teleport/legacy/types/types.proto [hunk 2]
```diff
@@ -505,6 +510,16 @@ message RedshiftServerless {
...
 
+// OpenSearch contains AWS OpenSearch specific metadata.
+message OpenSearch {
+  // DomainName is the name of the domain.
+  string DomainName = 1 [(gogoproto.jsontag) = "domain_name,omitempty"];
+  // DomainID is the ID of the domain.
+  string DomainID = 2 [(gogoproto.jsontag) = "domain_id,omitempty"];
+  // EndpointType is the type of the endpoint.
+  string EndpointType = 3 [(gogoproto.jsontag) = "endpoint_type,omitempty"];
+}
```
Comment (greedy52) [inline api/proto/teleport/legacy/types/types.proto:521]: Where are these values used? I don't mind keeping them but also seems unnecessary.

File: lib/srv/discovery/fetchers/db/aws_opensearch.go [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return nil, trace.Wrap(libcloudaws.ConvertRequestFailureError(err))
+	}
+
+	req := &opensearchservice.DescribeDomainsInput{DomainNames: []*string{}}
+	for _, domain := range names.DomainNames {
+		req.DomainNames = append(req.DomainNames, domain.DomainName)
+	}
+
+	domains, err := client.DescribeDomainsWithContext(ctx, req)
```
Comment (greedy52) [inline lib/srv/discovery/fetchers/db/aws_opensearch.go:145]: It is really unfortunate that names have to be listed first then fetch them. Sigh...

**Review by greedy52 — COMMENTED** [review]
Could we also update the db configurator or is it planned for another PR?

File: api/utils/aws/endpoint.go [hunk 4]
```diff
@@ -352,6 +352,13 @@ const (
 	MemoryDBClusterEndpoint = "cluster"
 	// MemoryDBNodeEndpoint is the endpoint of an individual MemoryDB node.
 	MemoryDBNodeEndpoint = "node"
+
+	// OpenSearchDefaultEndpoint is the default endpoint for domain.
+	OpenSearchDefaultEndpoint = "default"
+	// OpenSearchCustomEndpoint is the custom endpoint configured for domain.
+	OpenSearchCustomEndpoint = "custom"
```
[the remark this answers is not in the document]
Comment (Tener) [reply api/utils/aws/endpoint.go:359]: The use cases for this are likely niche, but the cost of implementing this was small, so I felt it was worth it.

File: lib/srv/discovery/fetchers/db/aws_opensearch.go (hunk 3 above)
Comment (Tener) [reply lib/srv/discovery/fetchers/db/aws_opensearch.go:145]: True, but OTOH, this isn't an interactive workload, so the extra call is not really a performance issue.

File: lib/services/database.go (hunk 1 above)
Comment (Tener) [reply lib/services/database.go:940]: Good call, I'll go with the latter.

File: api/proto/teleport/legacy/types/types.proto (hunk 2 above)
Comment (Tener) [reply api/proto/teleport/legacy/types/types.proto:521]: They are not used in the codebase right now, but these can be useful both in the future and for customer-side scripting.

File: lib/srv/discovery/fetchers/db/aws_opensearch.go (hunk 3 above)
Comment (marcoandredinis) [reply lib/srv/discovery/fetchers/db/aws_opensearch.go:145]:
```
// Container for the parameters to the DescribeDomains operation.
type DescribeDomainsInput struct {
	_ struct{} `type:"structure"`

	// Array of OpenSearch Service domain names that you want information about.
	// If you don't specify any domains, OpenSearch Service returns information
	// about all domains owned by the account.
	//
	// DomainNames is a required field
	DomainNames []*string `type:"list" required:"true"`
}
```

Do we really need the extra call?
> If you don't specify any domains, OpenSearch Service returns information about all domains owned by the account.

Isn't it what we want? all domains

Comment (Tener) [reply lib/srv/discovery/fetchers/db/aws_opensearch.go:145]: The docs appear incorrect in this case. 

If you comment out these lines:
```
	for _, domain := range names.DomainNames {
		req.DomainNames = append(req.DomainNames, domain.DomainName)
	}
```

The request returns an empty domain list, contrary to expectations.

Comment (marcoandredinis) [reply lib/srv/discovery/fetchers/db/aws_opensearch.go:145]: Indeed :+1: 
I also tested using the `aws` cli and got the same result: empty list

**Review by marcoandredinis — APPROVED** [review]
LGTM
I did a small test and it worked as expected 
image

**Review by smallinsky — APPROVED** [review]

**Pull request merged**